File: src/checkinprocessor.js
const redis = require('./utils/redisclient');
const logger = require('./utils/logger');
const sleep = require('./utils/sleep');

const runCheckinProcessor = async () => {
  const redisClient = redis.getClient();
  const checkinStreamKey = redis.getKeyName('checkins');
  const checkinProcessorIdKey = redis.getKeyName('checkinprocessor', 'lastid');
  const delay = process.argv[3] === 'delay';

  let lastIdRead = await redisClient.get(checkinProcessorIdKey);
  if (lastIdRead == null) {
    lastIdRead = 0;
  }

  logger.info(`Reading stream from last ID ${lastIdRead}.`);

  /* eslint-disable no-constant-condition */
  while (true) {
    /* eslint-enable */

    /* eslint-disable no-await-in-loop */
    const response = await redisClient.xread('COUNT', '1', 'BLOCK', '5000', 'STREAMS', checkinStreamKey, lastIdRead);
    /* eslint-enable */

    // If a stream entry was read, response looks like:
    // [
    //   [ "ncc:checkins",
    //     [
    //       [ "1609602085397-0",
    //         [ "locationId","171","userId","789","starRating","5" ]
    //       ]
    //     ]
    //   ]
    // ]

    if (response) {
      const checkinData = response[0][1][0];
      const fieldNamesAndValues = checkinData[1];

      const checkin = {
        id: checkinData[0],
        timestamp: checkinData[0].split('-')[0],
      };

      for (let n = 0; n < fieldNamesAndValues.length; n += 2) {
        const k = fieldNamesAndValues[n];
        const v = fieldNamesAndValues[n + 1];
        checkin[k] = v;
      }

      const userKey = redis.getKeyName('users', checkin.userId);
      const locationKey = redis.getKeyName('locations', checkin.locationId);

      logger.debug(`Updating user ${userKey} and location ${locationKey}.`);

      let pipeline = redisClient.pipeline();

      pipeline.hset(userKey, 'lastCheckin', checkin.timestamp, 'lastSeenAt', checkin.locationId);
      pipeline.hincrby(userKey, 'numCheckins', 1);
      pipeline.hincrby(locationKey, 'numCheckins', 1);
      pipeline.hincrby(locationKey, 'numStars', checkin.starRating);

      /* eslint-disable no-await-in-loop */
      const responses = await pipeline.exec();
      /* eslint-enable */

      // Calculate new averageStars... using the 3rd and 4th response
      // values from the pipeline (location numCheckins and location numStars).
      const locationNumCheckins = responses[2][1];
      const locationNumStars = responses[3][1];

      const newAverageStars = Math.round(locationNumStars / locationNumCheckins);
      lastIdRead = checkin.id;

      pipeline = redisClient.pipeline();
      pipeline.hset(locationKey, 'averageStars', newAverageStars);
      pipeline.set(checkinProcessorIdKey, lastIdRead);

      /* eslint-disable no-await-in-loop */
      await pipeline.exec();
      /* eslint-enable */

      logger.info(`Processed checkin ${checkin.id}.`);

      // Simulate some time consuming "work"...
      if (delay) {
        /* eslint-disable no-await-in-loop */
        await sleep.randomSleep(1, 10);
        /* eslint-enable */
      }
      /* eslint-enable */
    } else {
      logger.info('Waiting for more checkins...');
    }
  }
};

runCheckinProcessor();

File: src/scripts/checkinprocessor.lua
-- Example call from Redis
-- EVALSHA <sha> 2 user_hash_key location_hash_key timestamp location_id star_rating

local userKey = KEYS[1]
local locationKey = KEYS[2]
local checkinTimestamp = ARGV[1]
local checkinLocationId = ARGV[2]
local checkinStarRating = ARGV[3]

-- If the supplied timestamp is greater than the stored lastCheckin
-- timestamp, update it and the lastSeenAt field too.  If there isn't 
-- a current lastCheckin timestamp, use the value provided.
local currentLastCheckin = redis.call('hget', userKey, 'lastCheckin')

if ((currentLastCheckin == false or currentLastCheckin == nil)) or (tonumber(checkinTimestamp) > tonumber(currentLastCheckin)) then
    -- Update lastCheckin and lastSeenAt fields.
    redis.call('hset', userKey, 'lastCheckin', checkinTimestamp, 'lastSeenAt', checkinLocationId)
end

-- Increment the user's numCheckins and the location's numCheckins.
redis.call('hincrby', userKey, 'numCheckins', 1)
local locationNumCheckins = redis.call('hincrby', locationKey, 'numCheckins', 1)

-- Update the location's total star count.
local locationNumStars = redis.call('hincrby', locationKey, 'numStars', tonumber(checkinStarRating))

-- Calculate and store the location's new average star count.
local newAverageStars = math.floor((locationNumStars / locationNumCheckins) + 0.5)
redis.call('hset', locationKey, 'averageStars', tonumber(newAverageStars))
